sw/device/lib/runtime/pmp.c

```diff
@@ -0,0 +1,539 @@
+ *
+ * @param region PMP region to configure and set address for.
+ * @param address PMP address to be set.
+ * @return `pmp_region_configure_result_t`.
+ */
+pmp_region_configure_result_t pmp_csr_address_write(
+    pmp_region_index_t region, pmp_region_address_t address) {
+  pmp_region_address_t address_shifted = address >> PMP_ADDRESS_SHIFT;
+  if (!pmp_addr_csr_rw(region, kPmpCsrAccessTypeWrite, &address_shifted)) {
+    return kPmpRegionConfigureError;
+  }
+
+  pmp_region_address_t addr_csr_after_write;
+  if (!pmp_addr_csr_rw(region, kPmpCsrAccessTypeRead, &addr_csr_after_write)) {
+    return kPmpRegionConfigureError;
+  }
+
+  if (address_shifted != addr_csr_after_write) {
+    return kPmpRegionConfigureWarlError;
+  }
+
+  return kPmpRegionConfigureOk;
+}
+
+// TODO - short description.
+static bool pmp_cfg_permissions_set(pmp_region_permissions_t perm,
+                                    uint32_t *bitfield) {
+  switch (perm) {
+    case kPmpRegionPermissionsNone:
+      // No access is allowed.
+      break;
+    case kPmpRegionPermissionsReadOnly:
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_R, true);
+      break;
+    case kPmpRegionPermissionsExecuteOnly:
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_X, true);
+      break;
+    case kPmpRegionPermissionsReadExecute:
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_R, true);
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_X, true);
+      break;
+    case kPmpRegionPermissionsReadWrite:
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_R, true);
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_W, true);
+      break;
+    case kPmpRegionPermissionsReadWriteExecute:
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_R, true);
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_W, true);
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_X, true);
+      break;
+    default:
+      return false;
+  }
+
+  return true;
+}
+
+// TODO - short description.
+static void pmp_cfg_mode_lock_set(pmp_region_lock_t lock, uint32_t *bitfield) {
+  if (lock == kPmpRegionLockLocked) {
+    *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_L, true);
+  } else {
+    *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_L, false);
+  }
+}
+
+/**
+ * Check whether `address` is correctly aligned.
+ *
+ * The alignment depend on the granularity, which is implementation specific,
+ * and for Ibex is `PMP_GRANULARITY_IBEX`. Default granularity "G" is 0, which
+ * means a minimal alignment of 4bytes. Please see:
+ * "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+ * "3.6 Physical Memory Protection", "Figure 3.26" and section
+ * "Address Matching".
+ *
+ * @param address Conventional system address.
+ * @return `true` on success, `false` on failure.
+ */
+static bool pmp_address_aligned(pmp_region_address_t address) {
```

Comment:
Consider: `pmp_is_address_aligned`, to make it clear that this is a pure function that checks a property.

Though since you do other alignment checks in this file, maybe just having a single function that does alignment checks makes more sense...

Reply:
Done.